File: version_control_system/nit.py
import os
import sys

def find(a,b):
    # find(a,b) => bool
	result = a.lower().find(b.lower())
	if result == -1:
		return False
	else:
		return True



def sha1(filepath):
    # hash the contents of a file to get a unique identifier
    # from https://stackoverflow.com/questions/22058048/hashing-a-file-in-python
    
    import hashlib

    # BUF_SIZE is totally arbitrary, change for your app!
    BUF_SIZE = 65536  # lets read stuff in 64kb chunks!

    sha1 = hashlib.sha1()

    with open(filepath, 'rb') as f:
        while True:
            data = f.read(BUF_SIZE)
            if not data:
                break
            sha1.update(data)

    return sha1.hexdigest()


def date_time_string():
    import datetime
    return datetime.datetime.now().strftime("%Y%m%d_%H%M%S")


def init():
    for dir in [ ".nit", ".nit/files", ".nit/versions" ]:
        if not os.path.exists(dir): os.makedirs(dir)


def commit():
    top_dir = "."
    for root, dirs, files in os.walk(top_dir, topdown=False):

        # skip the .nit directory
        nit_directory = "/".join([top_dir, ".nit"])
        if root == nit_directory: continue

        version_file = "/".join([nit_directory, "versions", date_time_string() ])

        with open(version_file, "w") as output_file:
            for file in files: 
                path = "/".join([root, file])
                sha1_path = "/".join([nit_directory, "files", sha1(path) ])
                print(sha1_path, path)
                output_file.write(sha1_path +" "+ path +"\n")

# def revert( file ):
#     with open(filename) as f:
#         lines = f.read().split("\n")

#     for line in lines:

    

File: version_control_system/nit.r
rebol [
    doc: {
    === NIT VERSION CONTROL SYSTEM ===
    because I can't remember how git revert works

    The versions are stored in .nit/versions/<date-time>
    The files are stored in .nit/files/<sha1-hash> 
    }
]

file-walk: func [
	{Deep read a directory given a dir d
	an output block o and a boolean function fn}
	dir a_function /local file
] [
	file: read dir
	foreach file file [ do :a_function dir/:file ]
	foreach file file [ if #"/" = last file [ file-walk dir/:file :a_function ] ]
]

zero-extend: funct [ a n ] [ a: mold a  head insert/dup a "0" n - length? a ]

now-descending: funct [ ] [
	now-time: replace/all mold now/time ":" ""
	rejoin [ "" zero-extend now/year 4 zero-extend now/month 2 zero-extend now/day 2 "-" now-time ]
]

init: funct [ ] [

	attempt probe [
        make-dir %.nit/
        make-dir %.nit/files/
        make-dir %.nit/versions/
	]

]

commit: funct [ ] [

	version: copy [ ] 
	file-walk %. func [ filename ] [ 
	
		if find/match filename "./.nit" [ return ]
		
		filename_sha1: join %./.nit/files/ mold checksum/method filename 'sha1
		foreach c "#{}" [ replace/all filename_sha1 c "" ]
		write filename_sha1 read filename
		append version filename_sha1  
        append version filename   
	]   
	write/lines rejoin [ %.nit/versions/ now-descending ] version
	forskip version 2 [ print rejoin [ version/2 " => " version/1 ] ]
]


revert: funct [ version-file ] [

	version: read/lines version-file
	foreach f read %. [ if not find/match f ".nit" [ either dir? f [ delete-dir f ] [ delete f ] ] ]
	forskip version 2 [ print rejoin [ version/1 " => " version/2 ] attempt [ write to-file version/2 read to-file version/1 ] ]
]

